sklearn/utils/_array_api.py

```diff
@@ -74,22 +88,37 @@ def _check_array_api_dispatch(array_api_dispatch):
             )
 
 
-def device(x):
+def device(*array_list):
     """Hardware device the array data resides on.
 
     Parameters
     ----------
-    x : array
-        Array instance from NumPy or an array API compatible library.
+    *array_list : arrays
+        List of array instances from NumPy or an array API compatible library.
 
     Returns
     -------
     out : device
         `device` object (see the "Device Support" section of the array API spec).
     """
-    if isinstance(x, (numpy.ndarray, numpy.generic)):
-        return "cpu"
-    return x.device
+    if not array_list:
+        raise ValueError("At least one input array expected, got none.")
+
+    devices = set()
+    for array in array_list:
+        if isinstance(array, (numpy.ndarray, numpy.generic)) or not hasattr(
+            array, "device"
+        ):
+            devices.add("cpu")
+        else:
+            devices.add(array.device)
+
+    if len(devices) > 1:
+        raise ValueError(
+            f"Input arrays use different devices: {', '.join(sorted(devices))}"
+        )
```

Comment:
My bad. This would typically fail on non-str device attributes:

```python
In [1]: from sklearn.utils._array_api import device

In [2]: from array_api_compat import torch as xp

In [3]: a = xp.asarray([1], device="cpu")

In [4]: b = xp.asarray([1], device="mps")

In [5]: device(a, b)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 device(a, b)

File ~/code/scikit-learn/sklearn/utils/_array_api.py:118, in device(*array_list)
    114         devices.add(array.device)
    116 if len(devices) > 1:
    117     raise ValueError(
--> 118         f"Input arrays use different devices: {', '.join(sorted(devices))}"
    119     )
    121 return devices.pop()

TypeError: comparison not implemented
```

The following should fix it:

```suggestion
        device_names = ", ".join(sorted(str(d) for d in devices))
        raise ValueError(f"Input arrays use different devices: {device_names}")
```